### ISSUE TITLE
getVectorGrad returns transpose of gradient

### ISSUE THREAD

**Issue by bgranzow** [SCOREC/core#116]
The method ```apf::getVectorGrad``` appears to return the transpose of a vector gradient, rather than the gradient itself. To test this, I computed the gradient of a vector field myself, using nodal values and shape function gradients (wrt global coordinates), which I will refer to as method 1. Then I computed the gradient of a vector field with the method ```apf::getVectorGrad```, which I will refer to as method 2. These outputs of these methods return the transpose of one another. To demonstrate the methods, an example (missing some variable definitions) of this test is provided below:

I am very confident that method 1 produces the correct gradient, as I've compared the results to a widely used application code (Albany).

I've only verified that this problem occurs in 2D so far.

I do not currently have much time over the next two weeks to look into this, but hopefully this issue may be useful in the meantime.

```
apf::Matrix3x3 gradient(0,0,0,0,0,0,0,0,0);

// method 1
auto me = apf::getMeshElement(elem);
apf::getBF(shape, me, p, BF);
apf::getGradBF(shape, me, p, GBF);
for (int i = 0; i < num_dims; ++i) {
  for (int j = 0; j < num_dims; ++j) {
    gradient[i][j] = nodal(0, i) * GBF[0][j];
    for (int n = 1; n < num_nodes; ++n)
      gradient[i][j] += nodal(n, i) * GBF[n][j];
  }
}

// method 2
apf::getVectorGrad(elem, p, gradient);
```

**Comment by mortezah** [comment]
Assuming the standard definition for the gradient of a vector V(x)
(gradV)_{ij} = partial V_i / partial x_j ,
your observation is correct. Also for the Jacobian computation (which is basically gradient computation with respect to the parametric coordinates), apf routines return the transpose.

I can only speculate the reason behind this (at least in the case of the Jacobians). Sometimes users need to access the individual columns of the Jacobian (they correspond to tangent vectors which are used for normal computations, etc). If the apf routines return the **transpose of the Jacobian** then columns of the Jacobian can be easily accessed as follows

column_0 = J[0]
column_1 = J[1]
column_2 = J[2]

If, however, the routines where to return the actual Jacobian (not its transpose), getting the columns of the Jacobian would require a for-loop!

for i = 0,1,2
    column_0[i] = J[0][i]

**Comment by bgranzow** [comment]
Apparently everyone knew this already except for me...
I added some documentation with this commit 85c4584, which should be sufficient for this issue.

**Comment by bgranzow** [comment]
FYI: The previous documentation commit was put in the wrong place, which I fixed here: 0b4d33f
